File: source/skelet_Delphi/pisqpipe.pas
{ functions that communicate with manager through pipes }
{ modifications}
{  8.12.2008 - accessing undefined memory - marked ###F20081208 - Fontan}
unit pisqpipe;
interface
uses example; {change this to the name of your unit}

var
{ information about a game - you should use these variables }
width:integer=0;{ the board width }
height:integer; { the board height }
info_timeout_turn:integer=30000;  { time for one turn in milliseconds }
info_timeout_match:integer=1000000000; { total time for a game }
info_time_left:integer=1000000000;  { left time for a game }
info_max_memory:integer=0; { maximum memory in bytes, zero if unlimited }
info_game_type:integer=1;  { 0:human opponent, 1:AI opponent, 2:tournament, 3:network tournament }
info_exact5:integer=0; { 0:five or more stones win, 1:exactly five stones win }
info_renju:integer=0; { 0:Gomoku, 1:renju }
info_continuous:integer=0; { 0:single game, 1:continuous }
terminate:integer; { you must return from brain_turn when terminate>0 }
start_time:longword; { tick count at the beginning of turn }
dataFolder:string; { folder for persistent files }

{ these functions are implemented in this file }
procedure pipeOut(s:string);overload;
procedure pipeOut(fmt:string; args:array of const);overload;
procedure do_mymove(x,y:integer);
procedure suggest(x,y:integer);

implementation
uses windows,sysutils;
var cmd:string;
    tid:DWORD;
    event1,event2:THANDLE;

{* write a line to the pipe }
procedure pipeOut(s:string);overload;
begin
  writeln(s);
  Flush(output);
end;

procedure pipeOut(fmt:string; args:array of const);overload;
begin
  pipeOut(Format(fmt,args));
end;

{* parse two integers }
function parse_xy(param:string; var x,y:integer): boolean;
var i,ex,ey:integer;
begin
  i:=pos(',',param);
  val(copy(param,1,i-1),x,ex);
  val(copy(param,i+1,255),y,ey);
  if (i=0)or(ex<>0)or(ey<>0)or(x<0)or(y<0) then result:=false
  else result:=true;
end;

{* parse coordinates x,y }
function parse_coord(param:string; var x,y:integer): boolean;
begin
  result:=parse_xy(param,x,y);
  if (x>=width)or(y>=height) then result:=false
end;

{* parse coordinates x,y and player number z }
procedure parse_3int_chk(param:string; var x,y,z:integer);
var i,ex:integer;
    eyz:boolean;
begin
  i:=pos(',',param);
  val(copy(param,1,i-1),x,ex);
  eyz:=parse_xy(copy(param,i+1,255),y,z);
  if (i=0)or(ex<>0)or not eyz or(x<0)or(y<0)or(x>=width)or(y>=height) then z:=0;
end;

{* get word after cmd if input starts with cmd, otherwise return false }
function get_cmd_param(cmd,input:string; var param:string):boolean;
var n1,n2:integer;
begin
  n1:=length(cmd);
  n2:=length(input);
  if (n1>n2)or(StrLIComp(PChar(cmd),PChar(input),n1)<>0) then begin
    { it is not cmd }
    result:=false;
    exit;
  end;
  result:=true;
  repeat
    inc(n1);
    if n1>length(input) then begin {###F20081208}
      param:='';
      exit;
    end;
  until input[n1]<>' ';
  param:=copy(input,n1,255)
end;

{* send suggest }
procedure suggest(x,y:integer);
begin
  pipeOut('SUGGEST %d,%d',[x,y]);
end;

{* write move to the pipe and update internal data structures }
procedure do_mymove(x,y:integer);
begin
  brain_my(x,y);
  pipeOut('%d,%d',[x,y]);
end;

{* main function for the working thread }
function threadLoop(param:Pointer):DWORD;stdcall;
begin
  while true do begin
    WaitForSingleObject(event1,INFINITE);
    brain_turn();
    SetEvent(event2);
  end;
end;

{* start thinking }
procedure turn();
begin
  terminate:=0;
  ResetEvent(event2);
  SetEvent(event1);
end;

{* stop thinking }
procedure stop();
begin
  terminate:=1;
  WaitForSingleObject(event2,INFINITE);
end;

procedure start();
begin
  start_time:=GetTickCount();
  stop();
  if width=0 then begin
    width:=20;
    height:=20;
    brain_init();
  end;
end;

{* do command cmd }
procedure do_command();
var
  param,info,t:string;
  x,y,who,e:integer;
begin
  if get_cmd_param('info',cmd,param) then begin
    if get_cmd_param('max_memory',param,info) then info_max_memory:=StrToInt(info);
    if get_cmd_param('timeout_match',param,info) then info_timeout_match:=StrToInt(info);
    if get_cmd_param('timeout_turn',param,info) then info_timeout_turn:=StrToInt(info);
    if get_cmd_param('time_left',param,info) then info_time_left:=StrToInt(info);
    if get_cmd_param('game_type',param,info) then info_game_type:=StrToInt(info);
    if get_cmd_param('rule',param,info) then begin e:=StrToInt(info); info_exact5:=e and 1; info_continuous:=(e shr 1)and 1; info_renju:=(e shr 2)and 1; end;
    if get_cmd_param('folder',param,info) then dataFolder:=info;
{$IFDEF DEBUG_EVAL}
    if get_cmd_param('evaluate',param,info) then begin
      if parse_coord(info,x,y) then brain_eval(x,y);
    end;
{$ENDIF}
    { unknown info is ignored }
  end
  else if get_cmd_param('start',cmd,param) then begin
    width:= StrToInt(param);
    height:=width;
    start();
    brain_init();
  end
  else if get_cmd_param('rectstart',cmd,param) then begin
    if not parse_xy(param,width,height) then begin
      pipeOut('ERROR bad RECTSTART parameters');
    end else begin
      start();
      brain_init();
    end;
  end
  else if get_cmd_param('restart',cmd,param) then begin
    start();
    brain_restart();
  end
  else if get_cmd_param('turn',cmd,param) then begin
    start();
    if not parse_coord(param,x,y) then begin
      pipeOut('ERROR bad coordinates');
    end else begin
      brain_opponents(x,y);
      turn();
    end;
  end
  else if get_cmd_param('play',cmd,param) then begin
    start();
    if not parse_coord(param,x,y) then begin
      pipeOut('ERROR bad coordinates');
    end else begin
      do_mymove(x,y);
    end;
  end
  else if get_cmd_param('begin',cmd,param) then begin
    start();
    turn();
  end
  else if get_cmd_param('about',cmd,param) then begin
    pipeOut('%s',[infotext]);
  end
  else if get_cmd_param('end',cmd,param) then begin
    stop();
    brain_end();
    halt;
  end
  else if get_cmd_param('board',cmd,param) then begin
    start();
    while true do begin { fill the whole board }
      readln(cmd);
      parse_3int_chk(cmd,x,y,who);
      if who=1 then brain_my(x,y)
      else if who=2 then brain_opponents(x,y)
      else if who=3 then brain_block(x,y)
      else begin
        if StrIComp(PChar(cmd),'done')<>0 then pipeOut('ERROR x,y,who or DONE expected after BOARD');
        break;
      end;
    end;
    turn();
  end
  else if get_cmd_param('takeback',cmd,param) then begin
    start();
    t:='ERROR bad coordinates';
    if parse_coord(param,x,y) then begin
      e:= brain_takeback(x,y);
      if e=0 then t:='OK'
      else if e=1 then t:='UNKNOWN';
    end;
    pipeOut(t);
  end
  else begin
    pipeOut('UNKNOWN command');
  end;
end;


{* main function for AI console application  }
initialization
begin
  event1:=CreateEvent(nil,FALSE,FALSE,nil);
  CreateThread(nil,0,@threadLoop,nil,0,tid);
  event2:=CreateEvent(nil,TRUE,TRUE,nil);

  while true do begin
    readln(cmd);
    do_command();
  end;
end;
end.

File: source/skelet_CS/PisqPipe.cs
/** functions that communicate with Piskvork manager through pipes */
/** don't modify this file */
using System;
using System.Threading;

abstract class GomocupInterface
{
	/* information about a game - you should use these variables */
	public int width, height; /* the board size */
	public int info_timeout_turn = 30000; /* time for one turn in milliseconds */
	public int info_timeout_match = 1000000000; /* total time for a game */
	public int info_time_left = 1000000000; /* left time for a game */
	public int info_max_memory = 0; /* maximum memory in bytes, zero if unlimited */
	public int info_game_type = 1; /* 0:human opponent, 1:AI opponent, 2:tournament, 3:network tournament */
	public bool info_exact5 = false; /* false:five or more stones win, true:exactly five stones win */
	public bool info_renju = false; /* false:gomoku, true:renju */
	public bool info_continuous = false; /* false:single game, true:continuous */
	public int terminate; /* return from brain_turn when terminate>0 */
	public int start_time; /* tick count at the beginning of turn */
	public string dataFolder; /* folder for persistent files, can be null */

	/* you have to implement these functions */
	abstract public string brain_about { get; } /* copyright, version, homepage */
	abstract public void brain_init(); /* create the board and call Console.WriteLine("OK"); or Console.WriteLine("ERROR Maximal board size is .."); */
	abstract public void brain_restart(); /* delete old board, create new board, call Console.WriteLine("OK"); */
	abstract public void brain_turn(); /* choose your move and call do_mymove(x,y); 0<=x<width, 0<=y<height */
	abstract public void brain_my(int x, int y); /* put your move to the board */
	abstract public void brain_opponents(int x, int y); /* put opponent's move to the board */
	abstract public void brain_block(int x, int y); /* square [x,y] belongs to a winning line (when info_continuous is 1) */
	abstract public int brain_takeback(int x, int y); /* clear one square; return value: 0:success, 1:not supported, 2:error */
	abstract public void brain_end();  /* delete temporary files, free resources */
	virtual public void brain_eval(int x, int y) { } /* display evaluation of square [x,y] */


	private string cmd;
	private AutoResetEvent event1;
	private ManualResetEvent event2;

	/** read a line from STDIN */
	private void get_line()
	{
		cmd = Console.ReadLine();
		if (cmd == null) Environment.Exit(0);
	}

	/** parse coordinates x,y */
	private bool parse_coord2(string param, out int x, out int y)
	{
		string[] p = param.Split(',');
		if (p.Length == 2 && int.TryParse(p[0], out x) && int.TryParse(p[1], out y) && x >= 0 && y >= 0)
			return true;
		x = y = 0;
		return false;
	}

	/** parse coordinates x,y */
	private bool parse_coord(string param, out int x, out int y)
	{
		return parse_coord2(param, out x, out y) && x < width && y < height;
	}

	/** parse coordinates x,y and player number z */
	private void parse_3int_chk(string param, out int x, out int y, out int z)
	{
		string[] p = param.Split(',');
		if (!(p.Length == 3 && int.TryParse(p[0], out x) && int.TryParse(p[1], out y) && int.TryParse(p[2], out z)
			&& x >= 0 && y >= 0 && x < width && y < height))
			x = y = z = 0;
	}

	/** return pointer to word after command if input starts with command, otherwise return NULL */
	private static string get_cmd_param(string command, out string param)
	{
		param = "";
		int pos = command.IndexOf(' ');
		if (pos >= 0)
		{
			param = command.Substring(pos + 1).TrimStart(' ');
			command = command.Substring(0, pos);
		}
		return command.ToLower();
	}

	/** send suggest */
	protected void suggest(int x, int y)
	{
		Console.WriteLine("SUGGEST {0},{1}", x, y);
	}

	/** write move to the pipe and update internal data structures */
	protected void do_mymove(int x, int y)
	{
		brain_my(x, y);
		Console.WriteLine("{0},{1}", x, y);
	}

	/** main function for the working thread */
	private void threadLoop()
	{
		for (; ; )
		{
			event1.WaitOne();
			brain_turn();
			event2.Set();
		}
	}

	/** start thinking */
	private void turn()
	{
		terminate = 0;
		event2.Reset();
		event1.Set();
	}

	/** stop thinking */
	private void stop()
	{
		terminate = 1;
		event2.WaitOne();
	}

	private void start()
	{
		start_time = Environment.TickCount;
		stop();
		if (width == 0)
		{
			width = height = 20;
			brain_init();
		}
	}

	/** do command cmd */
	private void do_command()
	{
		string param, info;
		int x, y, who, e;

		switch (get_cmd_param(cmd, out param))
		{
			case "info":
				switch (get_cmd_param(param, out info))
				{
					case "max_memory":
						int.TryParse(info, out info_max_memory); break;
					case "timeout_match":
						int.TryParse(info, out info_timeout_match); break;
					case "timeout_turn":
						int.TryParse(info, out info_timeout_turn); break;
					case "time_left":
						int.TryParse(info, out info_time_left); break;
					case "game_type":
						int.TryParse(info, out info_game_type); break;
					case "rule":
						if (int.TryParse(info, out e))
						{
							info_exact5 = (e & 1) != 0;
							info_continuous = (e & 2) != 0;
							info_renju = (e & 4) != 0;
						}
						break;
					case "folder":
						dataFolder = info; break;
					case "evaluate":
						if (parse_coord(info, out x, out y)) brain_eval(x, y);
						break;
					/* unknown info is ignored */
				}
				break;
			case "start":
				if (!int.TryParse(param, out width) || width < 5)
				{
					width = 0;
					Console.WriteLine("ERROR bad START parameter");
				}
				else
				{
					height = width;
					start();
					brain_init();
				}
				break;
			case "rectstart":
				if (!parse_coord2(param, out width, out height) || width < 5 || height < 5)
				{
					width = height = 0;
					Console.WriteLine("ERROR bad RECTSTART parameters");
				}
				else
				{
					start();
					brain_init();
				}
				break;
			case "restart":
				start();
				brain_restart();
				break;
			case "turn":
				start();
				if (!parse_coord(param, out x, out y))
				{
					Console.WriteLine("ERROR bad coordinates");
				}
				else
				{
					brain_opponents(x, y);
					turn();
				}
				break;
			case "play":
				start();
				if (!parse_coord(param, out x, out y))
				{
					Console.WriteLine("ERROR bad coordinates");
				}
				else
				{
					do_mymove(x, y);
				}
				break;
			case "begin":
				start();
				turn();
				break;
			case "about":
				Console.WriteLine(brain_about);
				break;
			case "end":
				stop();
				brain_end();
				Environment.Exit(0);
				break;
			case "board":
				start();
				for (; ; ) /* fill the whole board */
				{
					get_line();
					parse_3int_chk(cmd, out x, out y, out who);
					if (who == 1) brain_my(x, y);
					else if (who == 2) brain_opponents(x, y);
					else if (who == 3) brain_block(x, y);
					else
					{
						if (!cmd.Equals("done", StringComparison.InvariantCultureIgnoreCase))
							Console.WriteLine("ERROR x,y,who or DONE expected after BOARD");
						break;
					}
				}
				turn();
				break;
			case "takeback":
				start();
				string t = "ERROR bad coordinates";
				if (parse_coord(param, out x, out y))
				{
					e = brain_takeback(x, y);
					if (e == 0) t = "OK";
					else if (e == 1) t = "UNKNOWN";
				}
				Console.WriteLine(t);
				break;
			default:
				Console.WriteLine("UNKNOWN command");
				break;
		}
	}


	/** main function for AI console application  */
	public void main()
	{
		try
		{
			int dummy = Console.WindowHeight;
			//ERROR, process started from the Explorer or command line
			Console.WriteLine("MESSAGE Gomoku AI should not be started directly. Please install gomoku manager (http://sourceforge.net/projects/piskvork). Then enter path to this exe file in players settings.");
		}
		catch (System.IO.IOException)
		{
			//OK, process started from the Piskvork manager
		}

		event1 = new AutoResetEvent(false);
		new Thread(threadLoop).Start();
		event2 = new ManualResetEvent(true);
		for (; ; )
		{
			get_line();
			do_command();
		}
	}

	static void Main(string[] args)
	{
		new GomocupEngine().main();
	}
}
